### ISSUE TITLE
Verify remote report from CLI

### ISSUE THREAD

**Issue by jumaffre** [openenclave/openenclave#2724]
In each Open Enclave install comes the `liboehostverify.a` library that can be linked to a binary if one needs to verify the remote report of an enclave (e.g. a client talking to a remote enclave running in ACC). 

However, this is impractical for clients that want to verify a remote report without having to build their own C++ binaries. For example, clients may communicate via off-the-shelf `curl` with a remote enclave running a HTTP server. 

I would be awesome if Open Enclave exported a binary to wrap the existing [`oe_verify_remote_report`](https://github.com/openenclave/openenclave/blob/master/include/openenclave/host_verify.h#L44) API so that clients can verify remote reports from the command line, e.g.:

```
$ oeverifyreport --report /path/to/remote/report
Report successfully verified
* OR *
Report verification failed: <OE_ERROR_CODE>
```

It is something that the CCF team has identified before (see https://github.com/microsoft/CCF/issues/683) but we strongly believe that contributing it directly to Open Enclave would also benefit other users of the SDK. 

We would be happy to help building this feature into Open Enclave if you're happy for us to do so. If yes, what would be the best place to build this feature (e.g. as a new [tool](https://github.com/openenclave/openenclave/tree/master/tools) or as an extension to `oecertdump` or `oesgx`)?

**Comment by jhand2** [comment]
I agree, this seems like a useful tool to have. Thanks for the suggestion!

I'm not sure I know the right place to put it. Is liboehostverify sgx-specific? If not, we could probably make a new tool.

**Comment by jazzybluesea** [comment]
How about creating general attestation CLI?  This cli would support generating evidence for any TEE and verify evidence + endorsements for any TEE?

**Comment by jhand2** [comment]
The reason I like a tool which just uses oehostverify is that it can be run on any machine, not just SGX. This is good for supporting the case @jumaffre mentioned. I just got remote evidence from some party and want to verify it with a tool. A CLI that also supports full attestation would require a hardware TEE on the machine.

Both may be useful but I think they serve different use-cases.

**Comment by achamayou** [comment]
It may be useful to have a tool that generates attestations as well, but as @jhand2 correctly points out this requires TEE hardware. The purpose here is to enable very lightweight verification for end users who talk to an in-enclave service, but don't necessarily have that capability.

I don't believe oehostverify is SGX-specific now, although it used to be (eg. #2264).

**Comment by jazzybluesea** [comment]
I think having a single tool minimizes the number of tools we have and also we have a single tool for attestation.  This single tool can check if it is running in compatible hardware before generating the report? and return an error message if it is not supported.

**Comment by achamayou** [comment]
In principle that’s fine, but it may be awkward for dependencies. oehostverify and its CLI could conceivably be shipped as a package with a very small dependency footprint (ie. no libsgx etc), separate from the full SDK.

If this is a single tool, then it has to have a number of optional dependencies. Optional dependencies are annoying because you can’t give an error if they’re missing, and so if they’re missing or incompatible by mistake, users find out late and with an error that often less helpful than “A requires B, you haven’t installed B”.

Also single large tools have larger —help, less immediate auto-complete... They aren’t always more user friendly :)

**Comment by yentsanglee** [comment]
A host verify sample that takes report as parameter can be helpful.
There's a commane-line test tool 'oecert' that generates report but it has to be built from OE_SDK.

**Comment by jumaffre** [comment]
Thanks all for your quick responses. It seems that there is consensus about the usefulness of such tool as part of the Open Enclave install. I agree with @jhand2 and @achamayou about having it as a very limited tool that does not require SGX (i.e. only link `liboehostverify.a`). 

I guess that the best place for it would be under https://github.com/openenclave/openenclave/tree/master/tools? I'm happy to have a go at it in the next few days if no one else is looking at it already.

**Comment by jhand2** [comment]
That would be great @jumaffre! Yes, the tools directory would be the best place for it. I'll go ahead and assign this issue to you, feel free to reach out with any questions.

**Comment by yentsanglee** [comment]
Host verification sample in PR 2865 as you pointed out is a light-weight verification tool. I think it's the tool/sample you are looking for.

**Comment by jhand2** [comment]
@jumaffre Does building and running that sample meet your needs?

**Comment by jumaffre** [comment]
That would be excellent! Do you know when this would be included in the hostverify package? Or does it come from free since #2865 is merged?

**Comment by jhand2** [comment]
Just saw your comment in #2865. It's included in the package already but it will just be the source. You would still need to compile it.

@yentsanglee @radhikaj Do either of you see a problem with installing a compiled version of the host_verify sample as a command-line tool in the hostverify deb package?

**Comment by jhand2** [comment]
If we install it I would like to include a printed disclaimer message saying that the command-line options are not a stable interface and are subject to change in future releases.

**Comment by radhikaj** [comment]
> Just saw your comment in #2865. It's included in the package already but it will just be the source. You would still need to compile it.
> 
> @yentsanglee @radhikaj Do either of you see a problem with installing a compiled version of the host_verify sample as a command-line tool in the hostverify deb package?

I have no issues and think it is a good idea.

**Comment by jumaffre** [comment]
@jhand2 @radhikaj I've implemented this in #3014.

**Comment by yentsanglee** [comment]
#2898 and #2948 are for publishing oecert which generates report and certificate. It's the symmetric tool of verify_host. @jumaffre, feel free to assign them to yourself.

**Comment by yentsanglee** [comment]
@jumaffre please close the issue if it's resolved.

**Comment by jhand2** [comment]
It is still in PR

**Comment by dthaler** [comment]
> However, this is impractical for clients that want to verify a remote report without having to build their own C++ binaries. For example, clients may communicate via off-the-shelf `curl` with a remote enclave running a HTTP server.
> 
> I would be awesome if Open Enclave exported a binary to wrap the existing [`oe_verify_remote_report`](https://github.com/openenclave/openenclave/blob/master/include/openenclave/host_verify.h#L44) API so that clients can verify remote reports from the command line, e.g.:
> 
> ```
> $ oeverifyreport --report /path/to/remote/report
> Report successfully verified
> * OR *
> Report verification failed: <OE_ERROR_CODE>
> ```

If the scenario is "clients that want to verify a remote report without having to build their own C++ binaries" then such a tool is not sufficient, since the example above assumes that you have a report (or evidence, more generally) in a file.   But how did it get there without building your own C++ binary?

**Comment by dthaler** [comment]
Reopening so that my question can be answered (I asked the same in the PR, but it was merged).  Can be closed if question can be answered in a way that does not warrant a documentation change.  If any documentation change is warranted, issue should remain open until such documentation is done.

**Comment by jumaffre** [comment]
@dthaler I will add extra documentation to the `host_verify` sample to clarify its use. However, now that this utility is installed as part of the SDK, I think this should also be documented in the `--help` message of the CLI. What do you think? I will issue a PR for this.

**Comment by jhand2** [comment]
Agreed, the tool should have a `-h|--help` option

**Comment by jumaffre** [comment]
Addressd in #3643
